utilities/rate_limiters/write_amp_based_rate_limiter.cc

```diff
@@ -0,0 +1,467 @@
+}
+
+void WriteAmpBasedRateLimiter::SetBytesPerSecond(int64_t bytes_per_second) {
+  assert(bytes_per_second > 0);
+  if (auto_tuned_.load(std::memory_order_acquire)) {
+    max_bytes_per_sec_.store(bytes_per_second, std::memory_order_relaxed);
+  } else {
+    SetActualBytesPerSecond(bytes_per_second);
+  }
+}
+
+void WriteAmpBasedRateLimiter::SetAutoTuned(bool auto_tuned) {
+  MutexLock g(&auto_tuned_mutex_);
+  if (auto_tuned_.load(std::memory_order_acquire) != auto_tuned) {
+    if (auto_tuned) {
+      max_bytes_per_sec_.store(rate_bytes_per_sec_, std::memory_order_relaxed);
+      refill_bytes_per_period_.store(
+          CalculateRefillBytesPerPeriod(rate_bytes_per_sec_),
+          std::memory_order_relaxed);
+    } else {
+      // must hold this lock to avoid tuner changing `rate_bytes_per_sec_`
+      MutexLock g2(&request_mutex_);
+      rate_bytes_per_sec_ = max_bytes_per_sec_.load(std::memory_order_relaxed);
+      refill_bytes_per_period_.store(
+          CalculateRefillBytesPerPeriod(rate_bytes_per_sec_),
+          std::memory_order_relaxed);
+    }
+    auto_tuned_.store(auto_tuned, std::memory_order_release);
+  }
+}
+
+void WriteAmpBasedRateLimiter::SetActualBytesPerSecond(
+    int64_t bytes_per_second) {
+  rate_bytes_per_sec_ = bytes_per_second;
+  refill_bytes_per_period_.store(
+      CalculateRefillBytesPerPeriod(bytes_per_second),
+      std::memory_order_relaxed);
+}
+
+bool WriteAmpBasedRateLimiter::IsFrontOfOneQueue(Req* r) {
+  request_mutex_.AssertHeld();
+  for (auto i = 0; i < Env::IO_TOTAL; ++i) {
+    if (!queue_[i].empty() && r == queue_[i].front()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void WriteAmpBasedRateLimiter::Request(int64_t bytes, const Env::IOPriority pri,
+                                       Statistics* stats) {
+  TEST_SYNC_POINT("WriteAmpBasedRateLimiter::Request");
+  TEST_SYNC_POINT_CALLBACK("WriteAmpBasedRateLimiter::Request:1",
+                           &rate_bytes_per_sec_);
+  if (auto_tuned_.load(std::memory_order_acquire) &&
+      (pri == Env::IO_HIGH || pri == Env::IO_USER) &&
+      duration_highpri_bytes_through_ + duration_bytes_through_ + bytes <=
+          max_bytes_per_sec_.load(std::memory_order_relaxed) * secs_per_tune_) {
+    // In the case where low-priority request is absent, actual time elapsed
+    // will be larger than secs_per_tune_, making the limit even tighter.
+    total_bytes_through_[pri] += bytes;
+    ++total_requests_[pri];
+    duration_highpri_bytes_through_ += bytes;
+    return;
+  }
+  assert(bytes <= refill_bytes_per_period_.load(std::memory_order_relaxed));
+  MutexLock g(&request_mutex_);
+
+  if (auto_tuned_.load(std::memory_order_acquire)) {
+    std::chrono::microseconds now(NowMicrosMonotonic(env_));
+    auto micros_per_tune = 1000 * 1000 * secs_per_tune_;
+    if (now - tuned_time_ >= std::chrono::microseconds(micros_per_tune)) {
+      Tune();
+    }
+  }
+
+  if (stop_) {
+    return;
+  }
+
+  ++total_requests_[pri];
```

Comment:
It looks like tikv ran into a segfault on this line. I'm still trying to understand why but I guess it's related to the addition of new IO priorities.

Here's the evidence:
1. segfault information:
```
[Mon Jun 24 10:59:14 2024] apply-1[2562783]: segfault at 7f68952f40a0 ip 000055bd6eacc09a sp 00007f6c1d621f90 error 6 in tikv-server[55bd69e00000+6052000]
```
2. The static address of the segfault code can be calculated as ip (000055bd6eacc09a) - base_addr_of_tikv (55bd69e00000) = 0x4ccc09a. 
3. With the help of gdb, we can locate the code of that address.
```
$ gdb tikv-server
(gdb) info line *0x4ccc09a
Line 172 of "/workspace/.cargo/git/checkouts/rust-rocksdb-9e01d192e8b6561d/af14652/librocksdb_sys/rocksdb/utilities/rate_limiters/write_amp_based_rate_limiter.cc"                                                                                               
   starts at address 0x4ccc093 <rocksdb::WriteAmpBasedRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)+147>                                                                                                                            
   and ends at 0x4ccc0a2 <rocksdb::WriteAmpBasedRateLimiter::Request(long, rocksdb::Env::IOPriority, rocksdb::Statistics*)+162>.
```

Reply:
Rocksdb introduced more types of priorities since 8.x, while WriteAmpRateLimiter only considered 3 of them. I thought it could work, but maybe this can cause some problems.

Reply:
One thing that seems to be interesting: the segfault issue seemed to only happen on Linux machines; I wasn't able to reproduce it on Mac. So it could be arch/compiler related. 
 
Also, I managed to get a coredump of the segfault on Linux. 
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00005595b5961efa in rocksdb::WriteAmpBasedRateLimiter::Request (this=0x7ffa1e214000, bytes=46541, pri=<optimized out>, stats=0x0)
    at /root/tabokie/packages/cargo/.cargo/git/checkouts/rust-rocksdb-da3ff04b5d606849/ca1f1dd/librocksdb_sys/rocksdb/utilities/rate_limiters/write_amp_based_rate_limiter.cc:172
172	  ++total_requests_[pri];
```
It shows that total_requests_ was initialized with a length of 4 as expected. But the `pri` was optimized out. 
```
(gdb) print total_requests_
$1 = {1407, 16, 274, 0}
(gdb) print pri
$2 = <optimized out>
```

Given the definition of IOPriority, the only way for it to cause a segfault is when `pri` equals IO_TOTAL, but I don't think that's how we expect `pri` to be used...
```
  enum IOPriority {
    IO_LOW = 0,
    IO_MID = 1,
    IO_HIGH = 2,
    IO_USER = 3,
    IO_TOTAL = 4
  };
```

Reply:
Found the problem! Turns out that one constructor for `Writer` did not initialize `rate_limiter_priority`. With this [one-line fix](https://github.com/hbisheng/rocksdb/commit/428b792cfcf04c50e63253b937da4cce266abbc8), the segfault problem went away. 

We might want to check how the bug was introduced and whether there could be other similar problems.

Reply:
👍